### ISSUE TITLE
SDN Controller application Features

### ISSUE THREAD

**Issue by u284976** [DSG-UniFE/ramp#2]
Hi 
I'm trying to use the feature of SDNController , but i have some JAVA exception occur, and my network is emulate by mininet's instruction `$sudo mn` like below:
```
node1 (10.0.0.1) is running SDNControllerService and SDNControllerClient both,
node2 (10.0.0.2) is running SDNControllerClient only.
```
When I send GenericDataType using "Send Packet" of node1's SDNClient to the node2, the following exception occurs
this is on h1
![send generic data on h1 display](https://user-images.githubusercontent.com/44360629/75873849-9dba1e80-5e4b-11ea-9a8b-257bd55d35ae.JPG)
this is on h2
![send generic data on h2 display](https://user-images.githubusercontent.com/44360629/75874390-8c254680-5e4c-11ea-96b1-d48f692e29c8.JPG)
Do you have similar mistakes?

And there are many terms that I don't understand, such as the difference between `GenericDataType`, `InfoDataType `and `DefaultMessage`
And what will the `data plane rule` affect?
Is There any related documents or papers that can make me understand?

Thanks so much!

**Comment by davidMonnuar** [comment]
Hello,

thank you for your feedback, I was able to repoduce the issue and now it is fixed. I've uploaded the latest version od Ramp Stand Alone Client in the deployment folder.

Answering your questions:
- Generic Data Type is a class that let to create a message with a payload and a sequence number
- Info Data Type is a class  that let to create a message with a payload, a sequence number and a value
- DefaultMessage is not used anymore
- Talking about the data plane rules, in the SDNControllerService there is a panel that lets you to activate/deactivate data plane rules for specific data types and you can also specify which nodes should activate such rules.

If you have more questions I'll be happy to answer, this time I'll be quicker.

Best regards,

Dmitrij

**Comment by u284976** [comment]
Hi,
thank you for your reply,
I tested it briefly, the sender's exception does not occur now, but some exception still occur. 
Is there anything I should notice to in the steps?

This time environment is the same with the previous, only the ID and sender changed
```
host1 node5 (10.0.0.1) is running SDNControllerService and SDNControllerClient both,
host2 node6 (10.0.0.2) is running SDNControllerClient only.
```
node 6 send a GenericDataType to node 5
node 6 is correct, but node5 had some exception like below
![send generic data on h1 display](https://user-images.githubusercontent.com/44360629/78295514-d8a09680-755e-11ea-8b78-9468fced1db1.JPG)


I don't know if I need to register with the controller, so I add Data Plane Rule with 
```
data type: GenericDataType 
Rule: DeserializationDataPlaneRule 
node5 
node6 

```

But more exception occurred....
node6 (sender)
![h2 after register data plane rule](https://user-images.githubusercontent.com/44360629/78296007-94fa5c80-755f-11ea-8cc9-8116bcab8175.JPG)

node5 (receiver)
![h1 after register data plane rule](https://user-images.githubusercontent.com/44360629/78296045-a2afe200-755f-11ea-8319-4e7a041422ab.JPG)


This is the video I recorded during the test. Could you please help me to see if there are steps that make mistakes?
https://youtu.be/yo7aqkVUfvk

Thanks so much!

**Comment by davidMonnuar** [comment]
Hello @u284976,

thanks for your feedback and for the video that helped me a lot to solve the issue quickly.

I've uploaded a new version of RampStandAloneClient in the deployment folder.

When you activate a rule in the SDNControllerService, if you directly click on the "Add Data Plane Rule" button it will automatically enables the rule for all the current SDNControllerClients so you can avoid to specify them. The select box that appears when you click the "Add client" button is useful only in case you want to activate the data plane rule for specific SDNControllerClients.

A last consideration, I've developed the SDNControllerService and SDNControllerClient during my thesis work so as you've noted is a little bit buggy since I didn't test it too much. So if you want to share what you would you like to do with Ramp I can help you in understanding the best approach to achieve it.

Have a nice day,

Dmitrij

**Comment by u284976** [comment]
Hi, @davidMonnuar 
Thanks for update, every function buttons are work correct now.

Actually I am a graduate student looking for the topic of graduation thesis. So far testing RAMP, I have three ideas.

The first is a whimsical idea, base on "Show Topology Graph" button in SDNControlService, I want to estimate client node's physical location by wireless link, example of situation like in a Smart Home (or other "less changeable spaces") , the signal of the network may not cover the entire range, and I can order the robot to go to the required location for temporary packet forward.

Second is base on [your past paper](https://ieeexplore.ieee.org/document/8390339), I want to explore "How to define a spontaneous network island range", in your paper that is a very important issue will affect the overall network QoS a lot, but it has not been resolved. However, the spontaneous network is very different from the real environment. If I want to do this topic, I think it is necessary to design a very large scenario, which is too difficult for me.

Third is also base on [your past paper](https://ieeexplore.ieee.org/document/6838365), this paper proposes a "task-> resource-> flow schedule" process to illustrate the architecture of SDN used in the IoT environment, on this basis, I want to add concept of "reservation", some un-important tasks can be adjusted by the controller to avoid network congestion.

For the third idea, I would like to ask SDNControlService can affect other Serivce?
For example, I press "Set Application Require" button with below figure parameter in node1 SDNControlClient, this setting can affect other Service? like FileSharingService bit rate will be limited to 50 kb/s.
![1](https://user-images.githubusercontent.com/44360629/78459485-d586e100-76eb-11ea-93a8-a1726b40fad4.JPG)

Thanks in advance!

**Comment by davidMonnuar** [comment]
Hi @u284976,

the first two points you mentioned are quite interesting.

About the third point we've already handled this case. In the SDNControllerService there is the possibility to specify many traffic engineering policies based on priority-based flows. These priorities can be set by specifying the ApplicationType value visibile in the screenshot you attached. Also the SDN Controller Client can specify which algorithm the Controller must use to compute a  path (i.e. the Minimum Network Load that selecets the path with the lowest actual network load).

Answering your question "For the third idea, I would like to ask SDNControlService can affect other Serivce?"

Keep in mind that RAMP in its base form is a distributed middleware using Dynamic Source Routing (DSR) to exchange messages between nodes. Also it supports a Discovery feature to publish and retrieve services offered by nodes inside the island. 
The SDNControllerService is an application built on top of RAMP Middleware using its API to act like a SDN Controller whose logic has been implemented on the application level and it is not a built-in feature provided by the middleware. It's a proof of concept of what RAMP is capable of in order to support higher-level functionalities like SDN.

For this reason SDNControllerService and SDNControllerClient are independent applications that work only the one with the other and what happens beetween them does not affect the other examples of RAMP applications which use the middleware in its base form.

In conclusion SDNController does not affect other services, if you desire this kind of behaviour you have to develop your own RAMP-SDN-compliant version of that service.

Hope I've helped

Dmitrij

**Comment by u284976** [comment]
Thank you for your help, let me understand the ability of RAMP and save me a lot of time.
I will try to coding the required function based on RAMP SDNControllerService.
Thank you again for your guidance and support!

**Comment by davidMonnuar** [comment]
Hi @u284976, You're welcome. 

Have a look at this link: https://github.com/DSG-UniFE/ramp/tree/master/src/test/sdncontroller

This is the test package containing several code examples of applications built on top of RAMP-SDN. In particular check the classes:
- ControllerClientTestSender
- ControllerClientTestReceiver

They are an example of communication between two SDN Clients inside the same island. The receiver registers itself as a RAMP service and the sender finds it via the RAMP Discovery Service. As soon on as the client finds the other one you can do whatever you want.

For the SDN Controller you have two options:
- Run the class ControllerServiceTest, it has no User Interface and you can change the SDN island configuration by code.
- Run the SDNControllerService, im my opinion the best option since you can control the island behaviour via a User Interface.

I think I've answered to all of your questions and I'm going to close this issue.

Have a nice day

Dmitrij
